internal/scm/github/scm.go

```diff
@@ -1,18 +1,599 @@
+	baseURL       string
+	ctx           context.Context
+	organizations []string
+}
+
+const (
+	perPageItemCount = 25
+)
+
+func NewSCM(options ...GithubOption) (scm *SCM, err error) {
+	scm = &SCM{}
+	for _, fn := range options {
+		if fn == nil {
+			continue
+		}
+		if err := fn(scm); err != nil {
+			return scm, err
+		}
+	}
+
+	ctx := context.Background()
+	scm.ctx = ctx
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: scm.token},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	c := github.NewClient(tc)
+
+	scm.client = c
+
+	return scm, nil
+}
+
+func (s *SCM) GetSelfOrganizations() (orgs []*github.Organization, err error) {
+	opt := &github.ListOptions{
+		PerPage: perPageItemCount,
+	}
+
+	organizations := []*github.Organization{}
+
+	for {
+		data, rsp, err := s.client.Organizations.List(s.ctx, "", opt)
+		if err != nil {
+			return organizations, err
+		}
+
+		organizations = append(organizations, data...)
+
+		if rsp.NextPage == 0 {
+			break
+		}
+
+		opt.Page = rsp.NextPage
+	}
+
+	return organizations, nil
+}
+
+func (s *SCM) listOrganizations() (orgs []*github.Organization, err error) {
+	if len(s.organizations) > 0 {
+		organizations := []*github.Organization{}
+		for _, orgName := range s.organizations {
+			organization, _, err := s.client.Organizations.Get(s.ctx, orgName)
+			if err != nil {
+				return organizations, err
+			}
+
+			organizations = append(organizations, organization)
+		}
+
+		return organizations, nil
+	}
+	return s.GetSelfOrganizations()
+}
+
+func (s *SCM) getPullRequest(owner, repo string, number int) (pr *github.PullRequest, err error) {
+	pr, _, err = s.client.PullRequests.Get(s.ctx, owner, repo, number)
+	return pr, err
+}
+
+func (s *SCM) OrganizationRepositoriesList(orgName string) (repos []*github.Repository, err error) {
+	opt := &github.RepositoryListByOrgOptions{
+		ListOptions: github.ListOptions{
+			PerPage: perPageItemCount,
+			Page:    1,
+		},
+	}
+
+	repos = []*github.Repository{}
+
+	for {
+		data, rsp, err := s.client.Repositories.ListByOrg(s.ctx, orgName, opt)
+		if err != nil {
+			return repos, err
+		}
+
+		repos = append(repos, data...)
+
+		if rsp.NextPage == 0 {
+			break
+		}
+
+		opt.Page = rsp.NextPage
+	}
+
+	return repos, nil
+}
+
+func (s *SCM) GetRepository(projectID int) (repository models.Repo, err error) {
+
+	repo, _, err := s.client.Repositories.GetByID(s.ctx, int64(projectID))
+	if err != nil {
+		return repository, err
+	}
+
+	repository = models.Repo{
+		ID:             projectID,
+		Name:           repo.GetName(),
+		Description:    repo.GetDescription(),
+		CreatorID:      int(repo.GetOwner().GetID()),
+		LastActivityAt: &repo.UpdatedAt.Time,
+	}
+
+	return repository, nil
+}
+
+func (s *SCM) ListMergeRequest(state, scope string, createdafterday int) (mergerequests models.MergeRequests, err error) {
+	mergerequests = []models.MergeRequest{}
+	filterdate := time.Now().AddDate(0, 0, -createdafterday)
+	mergedOnly := false
+
+	switch state {
```

Comment:
This switch case, to better testable and maintainabile may be extract to function.

Reply:
Done